### ISSUE TITLE
ODP.NET - OracleCommand.BindByName - Binding query parameters

### ISSUE THREAD

**Issue by AlphaGeek509** [FransBouma/Massive#252]
It appears that there is "feature", albeit I call it a bug, in the ODP.NET connector from Oracle.  It appears that when passing in parameters to an SQL query, the Named Parameters are ignored and the order by which the parameters are passed in takes precedence.  

The following websites confirm the existence as a bug.
- http://stackoverflow.com/questions/3876856/c-sharp-parameterized-queries-for-oracle-serious-dangerous-bug
- https://docs.oracle.com/html/E10927_01/OracleCommandClass.htm#DAFCJDDG
- http://stackoverflow.com/questions/3866275/why-am-i-getting-an-ora-01722-invalid-number
- http://stackoverflow.com/questions/1046632/binding-query-parameters-by-name-with-odp-net

**Solution**:  By all websites that I've read, they indicate that setting ODP.NET property of .BindByName to TRUE fixes the problem and allows proper binding.  My problem is I don't know where to place this property at  ;-)

Below is my sample code that I've ran to test the case:
-- Create table in Oracle

```
CREATE TABLE "PRINT_REV_CTRL" ("ID" NUMBER NOT NULL, "MFG_PART_ID" VARCHAR2(80BYTE));
```

-- Add constraint

```
ALTER TABLE "LT_DB"."PRINT_REV_CTRL" ADD CONSTRAINT "PRINT_REV_CTRL_ID_PK" PRIMARY KEY("ID");
```

--- Create Sequence

```
CREATE SEQUENCE  "LT_DB"."PRINT_REV_CTRL_SEQ"  MINVALUE 1100 MAXVALUE 99999999999999999999999999 INCREMENT BY 1 START WITH 1260 CACHE 20 NOORDER  NOCYCLE
```

``` c#
[Test]
        public void Update_SingleRow()
        {
            dynamic objPrc = new PrintRevControl();
            var inserted = objPrc.Insert(new { MFG_PART_ID = "TEST_PART_ID" });
            int insertedPrcID = inserted.ID;

            Assert.IsTrue(insertedPrcID > 0);
            inserted.MFG_PART_ID = "DI_TRAP_TSET";
            Assert.AreEqual(1, objPrc.Update(inserted, inserted.ID), "Update should have affected 1 row");

            var updatedRow = objPrc.Find(ID: inserted.ID);
            Assert.IsNotNull(updatedRow);
            Assert.AreEqual(inserted.ID, updatedRow.ID);
            Assert.AreEqual(inserted.MFG_PART_ID, updatedRow.MFG_PART_ID);
        }
```

The error will occur on objPrc.Update() command which eventually calls Execute in Massive.Shared.cs and fails on line 381.

**Comment by FransBouma** [comment]
Oh man, so true! I completely forgot about that. There's a solution, without needing to reference ODP.NET, which is to set the property through reflection. This can be done using a dynamic method, generated at runtime. I did this in LLBLGen Pro this way, using dynamic method compiler which generates IL at runtime: http://referencesource42.llblgen.com/#SD.LLBLGen.Pro.DQE.OracleODPNet/OracleSpecificCreator.cs,f805ba46d98e8fda,references This is a bit too much I think for Massive, we could cache the property descriptor though and simply set it on the command. It's a little slower than using IL but not that much, as it's the command, so not really noticeable.

**Comment by AlphaGeek509** [comment]
Please excuse my ignorance, but how and where is the best place to implement in the SetDbProviderFactoryParameterData method in Massive.Oracle.cs?  As I'm guess its needing to go there.

You mention IL.  What does IL refer to?

**Comment by AlphaGeek509** [comment]
Well, my version of the code is a hackfest, but it did the trick and parameterization worked for Oracle.  I'm now able to update in a mix-n-match fashion all I want.

I'll be anxiously awaiting for what the **PROPER** way to code this would be.

Thanks again Frans!!!

``` c#
private DbCommand CreateCommand(string sql, DbConnection conn, params object[] args)
        {
            var result = _factory.CreateCommand();
            /* Specific bug found in this code for Oracle specific database
             * Documented by these URLS:
             *  http://stackoverflow.com/questions/3876856/c-sharp-parameterized-queries-for-oracle-serious-dangerous-bug
             *  https://docs.oracle.com/html/E10927_01/OracleCommandClass.htm#DAFCJDDG
             */
            try
            {
                Type type = result.GetType();
                PropertyInfo prop = type.GetProperty("BindByName");
                prop.SetValue(result, true, null);
            }
            catch (Exception) { }

            if(result != null)
            {
                result.Connection = conn;
                result.CommandText = sql;
                result.AddParams(args);
            }
            return result;
        }
```

**Comment by FransBouma** [comment]
IL is the intermediate language the C# compiler compiles to, Jeff ;)

your code snippet is what I had in mind as well. It's not the fastest, but for commands and a library as Massive, which first and for all has to stay small, it's OK. Not sure when I have time for this though.

**Comment by FransBouma** [comment]
Implemented in pull #256
